### PR TITLE
712 - Add comments and event log to sidebar

### LINKED ISSUE #712: Submit/accept/reject - Show whether submitted node has been submitted/accepted/rejected
See mockups below for the outline that shows up (either green, orange, or red) to signify the status of their submitted node.

Add a new section to the sidebar called "Review". This section will be used to show the user the status of their submitted node and any associated comments. They will also be able to comment back here.

When a user submits a node, the author would see that node on their end only and there would be an orange line around the nodes that have been submitted and not accepted/rejected yet.

Potentially: User should see this as well for nodes that were accepted/rejected

### PR DESCRIPTION
**Pull request by narendrasss**
## Changes

Apologies for the huge PR - bear with me here 😄 

### Features

- **Node changes**
  - Change the edit button to a review button when a node is ready for review
    - When pressed, this button opens the sidebar to the review section
  - Update permissions check so reviewers don't have edit access to submitted draft nodes
- **Sidebar changes**
  - Add a review section to the sidebar that's visible for reviewers and node authors
    - This section shows a history of review-related actions performed on that node as well as a form to add comments and accept/reject a node (visible only to reviewers)
    - The review form in the node modal is now obsolete and has since been deleted.
  - Change the sidebar query url to show the active section e.g. `sidebar=review`
    - Previously the query url to open the sidebar was simply `sidebar=true`. This isn't sufficient because we want the ability to navigate directly to a specific section on the sidebar. 

### Internal Changes

- **New `/nodes/{nodeId}/review` endpoint**
  - Allow users to submit review events to a particular node via POST request.
    - A review "event" is currently either a comment or a change of status (e.g. submitted -> rejected). See [new schemas](#new-schemas) to see what each object looks like.
  - Added extensive data validation to ensure sent data is well formed.
  - **Rationale** - My first implementation attempt uses the PUT `/nodes/{nodeId}` endpoint. This failed when a subscriber wants to add a comment to a node they already submitted because the subscriber doesn't have edit access to that node anymore. This endpoint was created because there are different permissions in reviewing and updating a node.
- **Change `TapestryError` constructor signature**
  - New signature is `($code, $message = null, $status = null)`
  - **Rationale** - Typically adding new errors required adding a key to the `$ERRORS` array and setting an explicit message and status. This stopped me from adding descriptive error messages based on user input (e.g. I can't reply with "unknown property 'abc'). Backward-compatibility is ensured by checking if the code exists in the `$ERRORS` array first.

### New Schemas

Comment:

```
{
  type: "Comment"
  timestamp: string // in ISO8601 format
  comment: string
}
```

Status change:

```
{
  type: "StatusChange"
  timestamp: string // in ISO8601 format
  from: NodeStatus
  to: NodeStatus
}
```

## Screenshots

review_comment

review_accept

> Note this second GIF doesn't show the user's comments because it's a different node

## Issue Linkage

Closes #712 

## PR Dependency

Depends on: N/A

## Automated Testing

* Happy path of submitting a node -> accepting a node
* Happy path of submitting a node -> rejecting a node
* Checking if subscribers can submit a comment
* View different sections in sidebar
* View particular section in sidebar via url

### CODE REVIEW

**Comment by narendrasss** [comment]
> Small addition to scope: If the user tries to reject a node without leaving a comment, confirm with them that they want to reject without a comment. E.g. "Are you sure you want to reject this submission without a comment?"

**Comment by shirleywbi** [comment]
@narendrasss Still going through it but a couple of preliminary questions.

Are these buttons supposed to do anything aside from toggling open/close the sidebar?
image

Does the green ring stay indefinitely once accepted? Even to a public user?
image

Let's say a reviewer rejects a subscriber's node. The reviewers can no longer see the node. Do we need to handle the situation where the subscriber leaves a comment after rejection and thinks that the reviewer can see that comment (but actually can't because it's no longer visible to the reviewer)?

Subscribers can add a review node but can't edit their own review nodes until rejected. Is this expected behaviour?
image

**Comment by narendrasss** [comment]
@shirleywbi going to circle in @wynnset and @seanRong here since most of these questions are outside 712's scope.

> Are these buttons supposed to do anything aside from toggling open/close the sidebar?

It's supposed to open the sidebar to the respective section, e.g. clicking the review button will lead to the comment section.

> Does the green ring stay indefinitely once accepted? Even to a public user?

I believe so? This was part of #680 so maybe @seanRong or @wynnset can answer this.

> Let's say a reviewer rejects a subscriber's node. The reviewers can no longer see the node. Do we need to handle the situation where the subscriber leaves a comment after rejection and thinks that the reviewer can see that comment (but actually can't because it's no longer visible to the reviewer)?

I thought about this as well - I'm not so sure. We could always add a disclaimer saying your comments won't show until you resubmit the node. Thoughts @wynnset?

> Subscribers can add a review node but can't edit their own review nodes until rejected. Is this expected behaviour?

Yes - I think the idea was that you can't edit your homework once you submit it.

**Comment by seanRong** [comment]
> Does the green ring stay indefinitely once accepted? Even to a public user?

Yes, intended

**Comment by wynnset** [comment]
> > Let's say a reviewer rejects a subscriber's node. The reviewers can no longer see the node. Do we need to handle the situation where the subscriber leaves a comment after rejection and thinks that the reviewer can see that comment (but actually can't because it's no longer visible to the reviewer)?
Hmm.. I don't know why someone would leave a comment for a rejected node unless they want the reviewer to accept their node (e.g. saying that the reviewer's reason for rejecting was not valid)

File: templates/vue/src/components/node-review/ReviewLog.vue [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<template>
+  <ul>
+    <li v-for="event in eventWithComponents" :key="event.timestamp">
```
Comment (shirleywbi) [inline templates/vue/src/components/node-review/ReviewLog.vue:3]: I think the same key might be used for both the StatusChangeEvent and CommentEvent.
image
image

**Review by shirleywbi — COMMENTED** [review]
@narendrasss Looking good so far. Minor issue with one of the keys.

**Comment by wynnset** [comment]
Hey sorry @narendrasss the "[Re]Submit to Administrators for Review" button should also be moved to the sidebar. Sorry if I didn't understand the question raised by @shirleywbi, but I was under the impression that we were going to move the whole review process to the sidebar including submitting the node for approval. I don't think we'll need that disclaimer in that case either. The submitter should be able to leave a comment only if they are submitting again.

**Comment by wynnset** [comment]
I might make some minor visual edits once you have fixed that issue and it's been tested - I feel that the font sizes are a bit too all over the place. Also, the status log should not be as apparent (maybe lower opacity). Otherwise, it's looking great @narendrasss, great work!

**Comment by narendrasss** [comment]
@wynnset submit and resubmit have been moved to the sidebar now. Gonna hand it off to you to make the visual changes.

**Comment by wynnset** [comment]
Sorry @narendrasss do you mind fixing the merge conflicts since it's all your changes?

**Review by shirleywbi — COMMENTED** [review]
@narendrasss @wynnset
I think it's almost ready. Just a couple of comments/questions. 
1. Not sure if it's within the scope of this: Can submitted nodes be moved? Submitted nodes currently cannot be moved by anyone.
2. "Show me rejected nodes" in Tapestry Settings not working. When admin rejects a node, toggles the "Show me rejected nodes" in Tapestry Settings, and refreshes, the rejected node is not visible.
3. If two reviewers are reviewing at the same time, what is the expected behaviour if one accepts and the other rejects? Currently, it adds the node.

**Comment by wynnset** [comment]
@shirleywbi thanks for those points! @narendrasss these are my answers:

> 1. Not sure if it's within the scope of this: Can submitted nodes be moved? Submitted nodes currently cannot be moved by anyone.

This is not within the scope of this. There is a separate task to deal with that. 
> 2. "Show me rejected nodes" in Tapestry Settings not working. When admin rejects a node, toggles the "Show me rejected nodes" in Tapestry Settings, and refreshes, the rejected node is not visible.

This should get fixed
> 3. If two reviewers are reviewing at the same time, what is the expected behaviour if one accepts and the other rejects? Currently, it adds the node.

I think it's fine to leave it as is, but let's open a task in the backlog to change this so that an error is returned in the case of whoever loses that race condition.

File: templates/vue/src/components/TapestryNode.vue [hunk 2]
```diff
@@ -85,21 +85,32 @@
...
-          </node-button>
+          <template v-if="isLoggedIn">
+            <add-child-button
+              v-if="!isSubAccordionRow"
+              :node="node"
+              :x="canReview || hasPermission('edit') ? -35 : 0"
+              :y="radius"
+            ></add-child-button>
+            <node-button
+              v-if="canReview"
```
[the remark this answers is not in the document]
Comment (narendrasss) [reply templates/vue/src/components/TapestryNode.vue:96]: @wynnset can you confirm if this is something we want?

[the remark this answers is not in the document]
Comment (wynnset) [reply templates/vue/src/components/TapestryNode.vue:96]: This is a good idea, yes. Thanks!

**Comment by narendrasss** [comment]
> "Show me rejected nodes" in Tapestry Settings not working. When admin rejects a node, toggles the "Show me rejected nodes" in Tapestry Settings, and refreshes, the rejected node is not visible.

@shirleywbi I can't seem to replicate this:

https://user-images.githubusercontent.com/31267630/103811911-853e0600-5012-11eb-83d0-d3e392f39857.mov

**Comment by shirleywbi** [comment]
@narendrasss 
Hmm... I'm still having that issue as of commit 7ce88bb8306221b3cd7f6a31086c36ad6ce33ed5:
ezgif com-gif-maker

**Comment by narendrasss** [comment]
> @narendrasss
> Hmm... I'm still having that issue as of commit [7ce88bb](https://github.com/wynnset/tapestry-wp/commit/7ce88bb8306221b3cd7f6a31086c36ad6ce33ed5):
> ezgif com-gif-maker

Ah gotcha, I think I misunderstood what you said earlier. Should be fixed now!

**Review by wynnset — APPROVED** [review]

**Pull request merged**